{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import Bio.SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "def parse_location(location_str):\n",
    "    return location_str.split('..')\n",
    "\n",
    "def parse_string(location_str):\n",
    "    if location_str.index('complement') == 0:\n",
    "        return '-'\n",
    "    else:\n",
    "        return '+'\n",
    "\n",
    "def parse_entity(feature):\n",
    "    '''\n",
    "    Parse a transcribed entity\n",
    "    '''\n",
    "    entity = {}\n",
    "    # print('feature:', feature)\n",
    "    \n",
    "    entity['start'] = feature.location.start.position\n",
    "    entity['end'] = feature.location.end.position\n",
    "    entity['strand'] = '+' if feature.location.strand == 1 else '-'\n",
    "    #print(dir(feature))\n",
    "    #print(\"qulaifiers:\", feature.qualifiers)\n",
    "    if 'gene' in feature.qualifiers:\n",
    "        entity['name'] = feature.qualifiers['gene'][0]\n",
    "    \n",
    "    \n",
    "    for part in feature.location.parts:\n",
    "        entity['location_parts'] = {\n",
    "            'start': part.start.position,\n",
    "            'end': part.end.position,\n",
    "            'strand': '+' if part.strand == 1 else '-',\n",
    "        }\n",
    "\n",
    "    for xref in feature.qualifiers['db_xref']:\n",
    "        parts = xref.split(':')\n",
    "        if parts[0] == 'GeneID':\n",
    "            entity['geneId'] = parts[1]\n",
    "            break\n",
    "\n",
    "    if 'transcript_id' in feature.qualifiers:\n",
    "        entity['transcriptId'] = feature.qualifiers['transcript_id'][0]\n",
    "\n",
    "    #print('entity:', entity)\n",
    "    return entity\n",
    "\n",
    "def extract_refseq_features(refseq_file, sequence_handle=None):\n",
    "    '''\n",
    "    Extract a list of the features present in this file. This file should\n",
    "    only contain sequences for one record.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    refseq_file: string\n",
    "        The filename of a refseq file for a single sequence\n",
    "    sequence_name: string\n",
    "        The name of the sequence that these refseq annotations are for\n",
    "    sequence_handle File\n",
    "        A file handle for the sequence file\n",
    "        \n",
    "    Returns\n",
    "    --------\n",
    "    features: {}\n",
    "        A list of features indexed by GenBank gene ids\n",
    "    '''\n",
    "    record_count = 0\n",
    "    genes = {}\n",
    "    count = 0\n",
    "    \n",
    "    for record in Bio.SeqIO.parse(refseq_file, 'genbank'):\n",
    "        record_count += 1\n",
    "        #print(\"dir:\", dir(record))\n",
    "    \n",
    "        if sequence_handle is not None:\n",
    "            # store the sequence of this record in the provied fasta\n",
    "            # file handle\n",
    "            Bio.SeqIO.write(record, sequence_handle, 'fasta')\n",
    "        \n",
    "        #print('record:', record)\n",
    "        for feature in record.features:\n",
    "            # print(\"dir\", dir(feature))\n",
    "            # print(feature.type)\n",
    "            # print(feature.location, type(feature.location))\n",
    "            # print(dir(feature))\n",
    "            try:\n",
    "                curr_entity = parse_entity(feature)\n",
    "            except Exception as ex:\n",
    "                #print(\"Error parsing feature: {}\".format(feature), file=sys.stderr)\n",
    "                #print(\"Error: {}\".format(ex))\n",
    "                pass\n",
    "\n",
    "\n",
    "            curr_entity['type'] = feature.type\n",
    "            if 'geneId' in curr_entity:\n",
    "                #print('curr_entity:', curr_entity['geneId'])\n",
    "                pass\n",
    "\n",
    "            if feature.type == 'gene':\n",
    "                curr_entity['mRNAs'] = []\n",
    "                #print(\"feature:\", feature)\n",
    "                genes[curr_entity['geneId']] = curr_entity\n",
    "\n",
    "                if 'pseudo' in feature.qualifiers:\n",
    "                    curr_entity['type'] = 'pseudo'\n",
    "\n",
    "                # print(\"gene\", curr_entity)\n",
    "            elif feature.type == 'CDS':\n",
    "                #print(\"feature:\", feature)\n",
    "\n",
    "                cds = {}\n",
    "                cds['start'] = feature.location.start.position\n",
    "                cds['end'] = feature.location.end.position\n",
    "                genes[curr_entity['geneId']]['cds'] = cds\n",
    "            elif feature.type == 'ncRNA':\n",
    "                genes[curr_entity['geneId']] = curr_entity\n",
    "            elif feature.type == 'mRNA':\n",
    "                if 'mRNAs' not in genes[curr_entity['geneId']]:\n",
    "                    genes[curr_entity['geneId']]['mRNAs'] = []\n",
    "                genes[curr_entity['geneId']]['mRNAs'] += [curr_entity]\n",
    "            elif feature.type == 'misc_RNA':\n",
    "                genes[curr_entity['geneId']] = curr_entity        \n",
    "                #print(feature)\n",
    "            elif feature.type == 'precursor_RNA':\n",
    "                genes[curr_entity['geneId']] = curr_entity\n",
    "            elif feature.type == 'tRNA':\n",
    "                genes[curr_entity['geneId']] = curr_entity\n",
    "            else:\n",
    "                # print(feature)\n",
    "                pass\n",
    "\n",
    "            count += 1\n",
    "\n",
    "    print(\"len:\", len(genes))\n",
    "    return genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import os.path as op\n",
    "import urllib\n",
    "\n",
    "def extract_assembly_annotations(assembly_location):\n",
    "    #wget -O /tmp/assembly.txt ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.33_GRCh38.p7/GCF_000001405.33_GRCh38.p7_assembly_report.txt\n",
    "    !wget -O /tmp/assembly.txt {assembly_location}\n",
    "\n",
    "    !cat /tmp/assembly.txt | grep \"^#\" | tail -n 1 > /tmp/assembly_headers.txt\n",
    "    !cat /tmp/assembly.txt | grep -v \"^#\" > /tmp/assembly_processed.txt\n",
    "\n",
    "    with open('/tmp/assembly_headers.txt', 'r') as f:\n",
    "        headers = f.readlines()[0].strip().split()[1:]\n",
    "        print(headers)\n",
    "        \n",
    "    output_dir = op.join(op.expanduser('~/data/nuccore'),\n",
    "                         op.splitext(op.basename(assembly_location))[0])\n",
    "\n",
    "    if not op.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "\n",
    "    output_files = []\n",
    "    annotations = {\n",
    "        'sequences': []\n",
    "    }\n",
    "\n",
    "    with open('/tmp/assembly_processed.txt', 'r') as f:\n",
    "        output_fa = op.join(output_dir, 'seq.fa')\n",
    "        output_chromSizes = op.join(output_dir, 'chromSizes.tsv')\n",
    "        \n",
    "        if op.exists(output_fa):\n",
    "            os.remove(output_fa)\n",
    "        \n",
    "        if op.exists(output_chromSizes):\n",
    "            os.remove(output_chromSizes)\n",
    "            \n",
    "        f_chromsizes = open(op.join(output_dir, 'chromSizes.tsv'), 'a')\n",
    "            \n",
    "        for line in f:\n",
    "            parts = line.strip().split()\n",
    "            #print(\"parts:\", parts)\n",
    "            genbank_accn = parts[6]\n",
    "            sequence_name = parts[0]\n",
    "            ucsc_name = parts[-1]\n",
    "            refseq_name = parts[-5]\n",
    "            seq_len = int(parts[-2])\n",
    "            \n",
    "            common_name = ucsc_name if ucsc_name != 'na' else refseq_name\n",
    "            \n",
    "            f_chromsizes.write(\"{}\\t{}\\n\".format(common_name, seq_len))\n",
    "            #print('common_name:', common_name)\n",
    "\n",
    "            output_file  = op.join(output_dir, '{}.gb'.format(genbank_accn))\n",
    "            done_file = '{}.done'.format(output_file)\n",
    "            #print(output_file)\n",
    "\n",
    "            if not op.exists(done_file):\n",
    "                try:\n",
    "                    download_file = output_file + '.orig'\n",
    "                    url = 'https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id={}&rettype=gbwithparts&retmode=text'.format(genbank_accn)\n",
    "                    #print(\"url:\", url)\n",
    "                    urllib.request.urlretrieve(url, output_file)\n",
    "                    with open(done_file, 'w') as f1:\n",
    "                        pass\n",
    "                except Exception as ex:\n",
    "                    print('Ex:', ex)\n",
    "                    continue\n",
    "\n",
    "            #output_struct = dict(zip(headers, parts))\n",
    "\n",
    "            output_struct = {}\n",
    "            output_struct['refseqFilename'] = output_file\n",
    "            output_struct['sequence_name'] = sequence_name\n",
    "            output_struct['ucsc_name'] = ucsc_name\n",
    "            output_struct['refseq_name'] = refseq_name\n",
    "            output_struct['gene_features'] = extract_refseq_features(output_file,\n",
    "                                                                     sequence_handle=output_fa)\n",
    "            output_files += [output_struct]\n",
    "            annotations['sequences'] += [output_struct]\n",
    "        print(\"len(annotations)\", len(annotations['sequences']))\n",
    "\n",
    "    with open(op.join(output_dir, 'annotations.json'), 'w') as fo:\n",
    "        json.dump(annotations, fo, indent=2)\n",
    "    #print(output_files)\n",
    "    return output_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2018-04-30 22:56:35--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_assembly_report.txt\n",
      "           => ‘/tmp/assembly.txt’\n",
      "Resolving ftp.ncbi.nlm.nih.gov... 2607:f220:41e:250::12, 130.14.250.12\n",
      "Connecting to ftp.ncbi.nlm.nih.gov|2607:f220:41e:250::12|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/000/146/045/GCF_000146045.2_R64 ... done.\n",
      "==> SIZE GCF_000146045.2_R64_assembly_report.txt ... 2713\n",
      "==> EPSV ... done.    ==> RETR GCF_000146045.2_R64_assembly_report.txt ... done.\n",
      "Length: 2713 (2.6K) (unauthoritative)\n",
      "\n",
      "GCF_000146045.2_R64 100%[===================>]   2.65K  --.-KB/s    in 0s      \n",
      "\n",
      "2018-04-30 22:56:36 (6.53 MB/s) - ‘/tmp/assembly.txt’ saved [2713]\n",
      "\n",
      "['Sequence-Name', 'Sequence-Role', 'Assigned-Molecule', 'Assigned-Molecule-Location/Type', 'GenBank-Accn', 'Relationship', 'RefSeq-Accn', 'Assembly-Unit', 'Sequence-Length', 'UCSC-style-name']\n",
      "len: 101\n",
      "len: 432\n",
      "len: 184\n",
      "len: 799\n",
      "len: 317\n",
      "len: 143\n",
      "len: 585\n",
      "len: 305\n",
      "len: 232\n",
      "len: 392\n",
      "len: 338\n",
      "len: 572\n",
      "len: 505\n",
      "len: 418\n",
      "len: 579\n",
      "len: 497\n",
      "len: 46\n",
      "len(annotations) 1\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# assembly_location=\"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/215/GCF_000001215.4_Release_6_plus_ISO1_MT/GCF_000001215.4_Release_6_plus_ISO1_MT_assembly_report.txt\"\n",
    "# \n",
    "# assembly_location=\"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.37_GRCh38.p11/GCF_000001405.37_GRCh38.p11_assembly_report.txt\"\n",
    "\n",
    "# h37rv\n",
    "# assembly_location=\"ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/195/955/GCF_000195955.2_ASM19595v2/GCF_000195955.2_ASM19595v2_assembly_report.txt\"\n",
    "\n",
    "# sacCer3\n",
    "assembly_location='ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/146/045/GCF_000146045.2_R64/GCF_000146045.2_R64_assembly_report.txt'\n",
    "\n",
    "output_dir = extract_assembly_annotations(assembly_location)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def annotations_to_lines(annotations_json_file):\n",
    "    '''\n",
    "    Convert a list of annotations in JSON format to a text file.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    annotations_json_file: string\n",
    "        The filename containing the json file with the gene annotations\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "        A list of lines containing refGene formatted lines with the annotations\n",
    "    '''\n",
    "    with open(annotations_json_file, 'r') as f:\n",
    "        json_file = json.load(f)\n",
    "    \n",
    "        for sequence in json_file['sequences']:\n",
    "            common_name = sequence['ucsc_name'] if sequence['ucsc_name'] != 'na' else sequence['refseq_name']\n",
    "            print('common_name', common_name)\n",
    "            \n",
    "            for gene in sequence['gene_features'].values():\n",
    "                if 'cds' in gene:\n",
    "                    print(\"{chr}\\t{start}\\t{end}\".format(\n",
    "                        chr=common_name,\n",
    "                        start=gene['cds']['start'],\n",
    "                        end=gene['cds']['end']))\n",
    "                else:\n",
    "                    print\n",
    "                \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "common_name chrM\n",
      "chrM\t13817\t26701\n",
      "chrM\t13817\t23167\n",
      "chrM\t13817\t21935\n",
      "chrM\t13817\t19996\n",
      "chrM\t13817\t18830\n",
      "chrM\t13817\t16322\n",
      "chrM\t24155\t25255\n",
      "chrM\t27665\t27812\n",
      "chrM\t28486\t29266\n",
      "chrM\t36539\t43647\n",
      "chrM\t36539\t42251\n",
      "chrM\t36539\t40265\n",
      "chrM\t36539\t38579\n",
      "chrM\t46722\t46953\n",
      "chrM\t48900\t50097\n",
      "chrM\t61021\t61729\n",
      "chrM\t73757\t74513\n",
      "chrM\t74494\t75984\n",
      "chrM\t79212\t80022\n"
     ]
    }
   ],
   "source": [
    "lines = annotations_to_lines(op.join(output_dir, 'annotations.json'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/pete/data/nuccore/GCF_000146045.2_R64_assembly_report'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
